functions/Test-DbaCompression.ps1

```diff
@@ -0,0 +1,371 @@
+		$SqlCredential,
+		
+		[Alias("Databases")]
+		[object[]]
+		$Database,
+		
+		[object[]]
+		$ExcludeDatabase,
+		
+		[switch]
+		$IncludeSystemDBs,
+		
+		[switch]
+		$Silent
+	)
+	
+	begin {
+		Write-Message -Level System -Message "Bound parameters: $($PSBoundParameters.Keys -join ", ")"
+		$sql = "SET NOCOUNT ON;
+
+CREATE TABLE ##tmpCompression ([Schema] sysname
+	,[Table_Name] sysname
+	,[Index_Name] sysname NULL
+	,[Partition] int
+	,[Index_ID] int
+	,[Index_Type] VARCHAR(12)
+	,[Percent_Scan] smallint
+	,[Percent_Update] smallint
+	,[ROW_estimate_Pct_of_orig] bigint
+	,[PAGE_estimate_Pct_of_orig] bigint
+	,[Compression_Type_Recommendation] VARCHAR(7)
+	,size_cur bigint
+	,size_req bigint
+);
+
+CREATE TABLE ##tmpEstimateRow (
+	objname sysname
+	,schname sysname
+	,indid int
+	,partnr int
+	,size_cur bigint
+	,size_req bigint
+	,sample_cur bigint
+	,sample_req bigint
+);
+
+CREATE TABLE ##tmpEstimatePage (
+	objname sysname
+	,schname sysname
+	,indid int
+	,partnr int
+	,size_cur bigint
+	,size_req bigint
+	,sample_cur bigint
+	,sample_req bigint
+);
+
+INSERT INTO ##tmpCompression 
+([Schema]
+,[Table_Name]
+,[Index_Name]
+,[Partition]
+,[Index_ID]
+,[Index_Type]
+,[Percent_Scan]
+,[Percent_Update]
+)
+SELECT s.name AS [Schema], o.name AS [Table_Name], x.name AS [Index_Name],
+       i.partition_number AS [Partition], i.index_id AS [Index_ID], x.type_desc AS [Index_Type],
+       i.range_scan_count * 100.0 / (i.range_scan_count + i.leaf_insert_count + i.leaf_delete_count + i.leaf_update_count + i.leaf_page_merge_count + i.singleton_lookup_count) AS [Percent_Scan],
+       i.leaf_update_count * 100.0 / (i.range_scan_count + i.leaf_insert_count + i.leaf_delete_count + i.leaf_update_count + i.leaf_page_merge_count + i.singleton_lookup_count) AS [Percent_Update]
+FROM sys.dm_db_index_operational_stats (db_id(), NULL, NULL, NULL) i
+	INNER JOIN sys.objects o ON o.object_id = i.object_id
+	INNER JOIN sys.schemas s ON o.schema_id = s.schema_id
+	INNER JOIN sys.indexes x ON x.object_id = i.object_id AND x.index_id = i.index_id
+WHERE (i.range_scan_count + i.leaf_insert_count + i.leaf_delete_count + leaf_update_count + i.leaf_page_merge_count + i.singleton_lookup_count) <> 0
+	AND objectproperty(i.object_id,'IsUserTable') = 1
+ORDER BY [Table_Name] ASC;
+
+DECLARE @schema sysname, @tbname sysname, @ixid int
+DECLARE cur CURSOR FAST_FORWARD FOR SELECT [Schema], [Table_Name], [Index_ID] FROM ##tmpCompression
+OPEN cur
+FETCH NEXT FROM cur INTO @schema, @tbname, @ixid
+WHILE @@FETCH_STATUS = 0
+BEGIN
+	--SELECT @schema, @tbname
+	INSERT INTO ##tmpEstimateRow
+	(objname 
+	,schname 
+	,indid 
+	,partnr 
+	,size_cur 
+	,size_req 
+	,sample_cur 
+	,sample_req 
+	)
+	EXEC ('sp_estimate_data_compression_savings ''' + @schema + ''', ''' + @tbname + ''', ''' + @ixid + ''', NULL, ''ROW''' );
+	INSERT INTO ##tmpEstimatePage
+	EXEC ('sp_estimate_data_compression_savings ''' + @schema + ''', ''' + @tbname + ''', ''' + @ixid + ''', NULL, ''PAGE''');
```

Comment:
Both of these statements need to be changed to use `sp_executesql` instead of `EXEC`. You can see why [here](http://sqlblog.com/blogs/aaron_bertrand/archive/2011/09/17/bad-habits-to-kick-using-exec-instead-of-sp-executesql.aspx)